lib/cadet/incentives/achievements.ex

```diff
@@ -21,7 +21,43 @@ defmodule Cadet.Incentives.Achievements do
     |> Repo.all()
   end
 
-  @spec upsert(map()) :: {:ok, Achievement.t()} | {:error, {:bad_request, String.t()}}
+  @doc """
+  Returns all achievements with goals and progress.
+  # To Do
+  Run elixir tests for edge cases etc.
+
+  This returns Achievement structs with prerequisites, goal association and progress maps pre-loaded.
+  """
+  def get_total_xp(course_id) when is_ecto_id(course_id) do
+    achievements =
+      Achievement
+      |> where(course_id: ^course_id)
+      |> preload([:prerequisites, goals: [goal: :progress]])
+      |> Repo.all()
+
+
+    is_goal_completed = fn (goal_item) ->
+      if (Enum.count(goal_item.goal.progress) != 0),
+        do: Enum.at(goal_item.goal.progress, 0).completed,
+        else: false
+    end
+
+    is_all_goal_completed = fn (goal_item, acc) ->
+      if is_goal_completed.(goal_item), do: {:cont, acc}, else: {:halt, 0}
+    end
+
+    acc_goals = fn (all_goals) ->
+      Enum.reduce_while(all_goals, 1, is_all_goal_completed)
+    end
+
+    acc_achievements = fn (achievement_item, acc) ->
+      acc + acc_goals.(achievement_item.goals) * achievement_item.xp
+    end
+
+    Enum.reduce(achievements, 0, acc_achievements)
```

Comment:
Please don't declare helper functions like that, either use `defp` or indeed just inline the statements. These functions are small enough..

e.g.

```ex
    Enum.reduce(achievements, 0, fn (achievement_item, acc) ->

      completed_goals = Enum.reduce_while(achievement_item.goals, 1,
        fn (goal_item, acc) ->
          if Enum.count(goal_item.goal.progress) != 0 and Enum.at(goal_item.goal.progress, 0).completed,
            do: {:cont, acc},
            else: {:halt, 0}
        end)
      acc + completed_goals * achievement_item.xp
    end)
```

But this entire sequence is probably better written as:

```ex
    achievements
    |> Enum.map(fn ach -> 
        if Enum.all?(ach.goals, &(Enum.count(&1.goal.progress) != 0 and Enum.at(&1.goal.progress, 0).completed)),
          do: ach.xp,
          else: 0
      end)
    |> Enum.sum()
```

But then...

Your query at the top selects *all* goal progress entries in the course, for all users!

You need to pass in the course registration ID to this function, and adjust the query so that it only selects progress entries for the particular user you are calculating the XP for.

Finally.. I think this also doesn't handle achievements whose XP are dependent on the goal count.

Reply:
Most bugs here should be solved, still trying to comprehend the latter approach haha

Reply:
> still trying to comprehend the latter approach haha

In words:

1. For each achievement, take the achievement's XP award (`ach.xp`) if all goals are completed, else 0
2. Sum the numbers in the list resulting from step 1